frontend/catalyst/jax_tracer.py

```diff
@@ -106,7 +106,7 @@ def _eval_jaxpr(*args):
 
 KNOWN_NAMED_OBS = (qml.Identity, qml.PauliX, qml.PauliY, qml.PauliZ, qml.Hadamard)
 
-FORCED_ORDER_PRIMITIVES = {qdevice_p, qextract_p, qinst_p}
+FORCED_ORDER_PRIMITIVES = {qdevice_p}
```

Comment:
I have a question here: Does the primitive ordering work when there is only one primitive in the set? I.e. if we only have `qdevice_p` in the set, what is the order forced with respect to?

The other comment is that we should add the print primitive here, that might fix one of our open issues (#407 )

Reply:
> I have a question here: Does the primitive ordering work when there is only one primitive in the set? I.e. if we only have `qdevice_p` in the set, what is the order forced with respect to?

Good question! The way `sort_eqns()` is currently implemented, each equation with a forced order primitive will introduce a constraint to _all_ equations following it (regardless of whether or not those equations also contain a forced order primitive). I can't be certain if this is intentional.

> The other comment is that we should add the print primitive here, that might fix one of our open issues (#407 )

Let's give it a shot!

Reply:
@dime10 @erick-xanadu It doesn't look like adding the `print_p` primitive to the set fixes #407 because the `print` statements are captured during the classical trace and there are no data dependencies which can be used to relate the `print` statements to the quantum operations during the sorting phase.

Reply:
> Good question! The way sort_eqns() is currently implemented, each equation with a forced order primitive will introduce a constraint to all equations following it (regardless of whether or not those equations also contain a forced order primitive). I can't be certain if this is intentional.

Got it, in that case it should work as intended with just the `qdevice_p` primitive in the set :) 

> @dime10 @erick-xanadu It doesn't look like adding the print_p primitive to the set fixes https://github.com/PennyLaneAI/catalyst/issues/407 because the print statements are captured during the classical trace and there are no data dependencies which can be used to relate the print statements to the quantum operations during the sorting phase.

@Mandrenkov Hmm, are you saying the sorting doesn't consider equations that were captured during the classical tracing part? (no data dependencies is expected, similar to the qdevice primitive)

> As a note, I am not opposed to keeping issue https://github.com/PennyLaneAI/catalyst/issues/407 open if we can find an alternative way. If we can remove this sort, I would be happy!

@erick-xanadu Some part of the sorting is definitely needed, because the classical and quantum equations need to be merged together in a manner that obeys the dataflow. But the side effect sorting we have (forced order primitives) could probably be replaced with the effects mechanism in JAX. I think that's the "official" way of doing this. (Definitely out of scope here though.)

Reply:
> @erick-xanadu Some part of the sorting is definitely needed, because the classical and quantum equations need to be merged together in a manner that obeys the dataflow. But the side effect sorting we have (forced order primitives) could probably be replaced with the effects mechanism in JAX. I think that's the "official" way of doing this. (Definitely out of scope here though.)

I agree with out of scope. But I think there's a way to get rid of this sorting (potentially sacrificing some of the integration with PL) The basic idea is to have the operations be applied instead of queued when doing tracing. Then we don't have to manually bind them after the classical part has been bound nor sort them after binding. This would order the quantum operations in the order that we want. I don't believe JAX reorders operations at this stage. They should be scheduled as they are bound.

Reply:
Yes that's something we discussed many times (eliminating the queuing part altogether), but we abandoned the idea in order to support tape transforms. Additionally, there was always the issue that certain already recorded operations would need to be removed again from the jaxpr in a post-processing step, due to how PL constructs an operation first before it is being passed to another operation that wraps the former.

Additionally, @rmoyard had an idea recently to actually preserve (some of) the separation between classical and quantum instructions, rather then merging them in an indiscriminate fashion. I think the idea deserves some exploration to see what benefits we could draw from it.

Reply:
>@Mandrenkov Hmm, are you saying the sorting doesn't consider equations that were captured during the classical tracing part? (no data dependencies is expected, similar to the qdevice primitive)

@dime10 Not quite. For example, if you were printing the output of some classical operation, then you can be sure that the equations will be ordered in such a way that the classical value is computed before it is printed. This is similar to the `qdevice_p` case except that `qdevice_p` (AFAIK) never has any data dependencies, whereas the `print()` one conceivably could.

> The basic idea is to have the operations be applied instead of queued when doing tracing. Then we don't have to manually bind them after the classical part has been bound nor sort them after binding. This would order the quantum operations in the order that we want.

@erick-xanadu Indeed, that is something I'm experimenting with now! @dime10's point about tape transforms is a good one, although it may not be a dealbreaker if switching away from tapes would make certain (important) things now possible to do.

Reply:
Thanks for the discussion, from my side I'm going to close this thread to unblock the PR :)